### PR TITLE
Feature/adding struct log

### PR DESCRIPTION
**Pull request by rhiGriff**
**What is the context of this PR?**

Link to Trello card
https://trello.com/c/0Fn8SNCC 

Describe what you have changed and why, link to other PRs or Issues as appropriate.\

**How to review**

Any notes for the reviewer  (include screenshots if appropriate).

### CODE REVIEW

File: app/repository/retriever.py [hunk 1]
```diff
@@ -142,10 +143,11 @@ def retrieve_thread(thread_id, user, _survey=constants.BRES_SURVEY):
                                else_=0).desc(), Events.event.desc())
 
             if result.count() == 0:
+                logger.debug('Tread does not exist', thread_id=thread_id)
```
Comment (emilioward) [inline app/repository/retriever.py]: Thread does not exist*

File: app/validation/domain.py [hunk 2]
```diff
@@ -70,17 +71,18 @@ def validate_to(self, msg_to):
...
+                logger.debug('Not a valid user', user=item)
                 raise ValidationError("{0} is not a valid user.".format(item))
 
     @validates("msg_from")
     def validate_from(self, msg_from):
         self.validate_non_zero_field_length("msg_from", len(msg_from), constants.MAX_FROM_LEN)
         if g.user.is_internal and msg_from != constants.BRES_USER:
-            raise ValidationError('You are not authorised to send a message on behalf of user or work group {0}'
-                                  .format(msg_from))
+            logger.debug('Not authorised to send a message on behalf of user or work group', fessage_from=msg_from)
```
Comment (emilioward) [inline app/validation/domain.py]: message from not `fessage`

File: app/validation/domain.py [hunk 3]
```diff
@@ -70,17 +71,18 @@ def validate_to(self, msg_to):
...
+            raise ValidationError('You are not authorised to send a message on behalf of user or work group {0}'.format(msg_from))
         if g.user.is_respondent and msg_from != g.user.user_uuid:
-            raise ValidationError('You are not authorised to send a message on behalf of user or work group {0}'
-                                  .format(msg_from))
+            logger.debug('Not authorised to send a message on behalf of user or work group', fessage_from=msg_from)
```
Comment (emilioward) [inline app/validation/domain.py]: Same again

File: app/validation/domain.py [hunk 4]
```diff
@@ -167,19 +174,20 @@ def validate_to(self, msg_to):
...
-            logger.debug("msg_to field empty")
+            logger.debug('msg_to field empty')
 
     @validates("msg_from")
     def validate_from(self, msg_from):
         self.validate_field_length(msg_from, len(msg_from), constants.MAX_FROM_LEN)
         if g.user.is_internal and msg_from != constants.BRES_USER:
-            raise ValidationError('You are not authorised to save a draft on behalf of user or work group {0}'
-                                  .format(msg_from))
+            logger.debug('Not authorised to save a draft on behalf of user or work group', fessage_from=msg_from)
```
Comment (emilioward) [inline app/validation/domain.py]: Another spelling error here and one more below

File: app/services/party_service.py [hunk 5]
```diff
@@ -15,6 +16,7 @@ def get_business_details(ru):
         """Retrieves the business details from the party service"""
         url = app.settings.RAS_PARTY_GET_BY_BUSINESS.format(app.settings.RAS_PARTY_SERVICE, ru)
         party_data = requests.get(url, verify=False)
+
         logger.debug('party get business details result => {} {} : {}'.format(party_data.status_code, party_data.reason,
```
Comment (rjweeks) [inline app/services/party_service.py]: These log statements need updating?

File: app/services/party_service.py [hunk 6]
```diff
@@ -43,6 +45,7 @@ def get_user_details(uuid):
             else:
                 url = app.settings.RAS_PARTY_GET_BY_RESPONDENT.format(app.settings.RAS_PARTY_SERVICE, uuid)
                 party_data = requests.get(url, verify=False)
+
                 logger.debug('party get user details result => {} {} : {}'.format(party_data.status_code,
```
Comment (rjweeks) [inline app/services/party_service.py]: and this one?

File: app/api_mocks/party_service_mock.py [hunk 7]
```diff
@@ -22,7 +23,7 @@ def get_user_details(self, uuid):
         try:
             return Response(response=json.dumps(self._respondent_ids[uuid]), status=200, mimetype="text/html")
         except KeyError:
-            logger.debug('User ID %s not in mock party service', uuid)
+            logger.debug('User ID not in mock party service', uuid=uuid)
```
Comment (necrophonic) [inline app/api_mocks/party_service_mock.py]: Would `user_id` be a better key in the message for context? (Or `user_uuid` to be specific)

File: app/application.py [hunk 8]
```diff
@@ -74,7 +74,7 @@ def before_request():
         log_request()
         res = authenticate(request.headers)
         if res != {'status': "ok"}:
-            logger.debug("Failed to authenticate user")
+            logger.debug('Failed to authenticate user')
```
Comment (necrophonic) [inline app/application.py]: Which user?

File: app/application.py [hunk 9]
```diff
@@ -115,4 +115,4 @@ def log_request():
         args_list.append('arg ' + str(count) + ' = ' + str(key) + ': ' + str(val))
 
     params = ''.join(args_list)
-    logger.debug('Headers: ' + str(headers) + ' Body: ' + str(req_data) + ' Arguments: ' + str(params))
+    logger.debug('Incoming request', headers=headers, req_data=req_data, arguments=params)
```
Comment (necrophonic) [inline app/application.py]: Being that this is secure messaging, is there anything (personally identifying info etc) in the `headers`, `req_data` or `params`?

File: app/authentication/authenticator.py [hunk 10]
```diff
@@ -21,18 +22,18 @@ def authenticate(headers):
 
     else:
         res = Response(response="Authorization header required.", status=400, mimetype="text/html")
-        logger.debug("Authorization header not supplied.")
+        logger.debug('Authorization header not supplied.')
         return res
```
Comment (necrophonic) [inline app/authentication/authenticator.py]: Could clean this up and do it with less assignment:
```
logger.info("Authorization header not supplied")
return Response(response="Authorization header required.", status=400, mimetype="text/html")
```

File: app/authentication/authenticator.py [hunk 11]
```diff
@@ -21,18 +22,18 @@ def authenticate(headers):
...
 
 
 def check_jwt(token):
     JWT_ENCRYPT = settings.SM_JWT_ENCRYPT
-    logger.debug('JWT Encryption (0=disabled, 1=enabled) : {}'.format(JWT_ENCRYPT))
+    logger.debug('JWT Encryption (0=disabled, 1=enabled)', JWT_ENCRYPT=JWT_ENCRYPT)
```
Comment (necrophonic) [inline app/authentication/authenticator.py]: Surely should always be enabled?

File: app/authentication/authenticator.py [hunk 12]
```diff
@@ -21,18 +22,18 @@ def authenticate(headers):
...
     try:
         if JWT_ENCRYPT == '1':
             decrypter = Decrypter()
             decrypted_jwt_token = decrypter.decrypt_token(token)
-            logger.debug("Decrypted JWT.")
+            logger.debug('Decrypted JWT.')
```
Comment (necrophonic) [inline app/authentication/authenticator.py]: Doesn't give any context. Also we don't normally finished messages with a period '.'

File: app/connector.py [hunk 13]
```diff
@@ -25,11 +25,11 @@ def get_database_uri():
         return os.environ.get('SECURE_MESSAGING_DATABASE_URL', 'sqlite:////tmp/messages.db')
 
     for key, value in decoded_config.items():
-        logger.info('Inspecting key: "' + str(key) + '" with value: ' + str(value))
+        logger.info('Inspecting key and value', key=key, value=value)
```
Comment (necrophonic) [inline app/connector.py]: No context - also are these things that should be in the log?

File: app/connector.py [hunk 14]
```diff
@@ -25,11 +25,11 @@ def get_database_uri():
...
         if decoded_config[key][0]['name'] == 'secure-message-db':
             creds = decoded_config[key][0]['credentials']
             uri = creds['uri']
-            logger.info('Postgres DATABASE URI: ' + uri)
+            logger.info('Postgres DATABASE URI:', uri=uri)
```
Comment (necrophonic) [inline app/connector.py]: Shouldn't be logging `uri`s

File: app/repository/modifier.py [hunk 15]
```diff
@@ -24,7 +24,7 @@ def add_label(label, message, user, session=db.session):
             return True
         except Exception as e:
             session.rollback()
-            logger.error(e)
+            logger.error('Error retrieving messages from database', error=e)
```
Comment (necrophonic) [inline app/repository/modifier.py]: More context about what it's doing (as the same message seems to be fired from multiple places?).

File: app/resources/drafts.py [hunk 16]
```diff
@@ -23,6 +23,7 @@ def post(self):
         draft = DraftSchema().load(post_data)
 
         if 'msg_id' in post_data:
+            logger.debug('Message cannot include message ID')
```
Comment (necrophonic) [inline app/resources/drafts.py]: `error` rather than `debug`

File: app/resources/drafts.py [hunk 17]
```diff
@@ -37,6 +38,7 @@ def post(self):
         else:
             res = jsonify(draft.errors)
             res.status_code = 400
+            logger.debug('Failed saving draft', status_code=res.status_code)
```
Comment (necrophonic) [inline app/resources/drafts.py]: `error` rather than `debug`

File: app/resources/drafts.py [hunk 18]
```diff
@@ -96,18 +98,20 @@ def put(self, draft_id):
         """Handles modifying of drafts"""
         data = request.get_json()
         if 'msg_id' not in data:
+            logger.debug('Draft put requires message ID')
```
Comment (necrophonic) [inline app/resources/drafts.py]: Anything that's raising an exception should be an `error` instead of a `debug`

File: app/resources/messages.py [hunk 19]
```diff
@@ -81,6 +82,7 @@ def post(self):
         else:
             resp = jsonify(message.errors)
             resp.status_code = 400
+            # add debug for error
```
Comment (necrophonic) [inline app/resources/messages.py]: Add the logging or remove the comment

File: app/resources/messages.py [hunk 20]
```diff
@@ -132,8 +134,10 @@ def get(message_id):
         else:
             result = jsonify({'status': 'error'})
             result.status_code = 403
+            logger.debug('Error getting message by ID', status_code=result.status_code)
```
Comment (necrophonic) [inline app/resources/messages.py]: `error`

File: app/resources/messages.py [hunk 21]
```diff
@@ -157,6 +161,7 @@ def put(message_id):
         else:
             res = jsonify({'status': 'error'})
             res.status_code = 400
+            logger.debug('Error updating message', status_code=res.status_code)
```
Comment (necrophonic) [inline app/resources/messages.py]: `error`

File: app/application.py (hunk 9 above)
Comment (andrewmil) [inline app/application.py]: We send the JWT in the headers, probably should be doing that?

File: app/api_mocks/party_service_mock.py (hunk 7 above)
Comment (andrewmil) [inline app/api_mocks/party_service_mock.py]: We should make sure whatever we're calling this (and other things) are consistent across services as well. I know in frontstage we're call this party_id most of the time

File: app/validation/domain.py [hunk 22]
```diff
@@ -119,20 +121,22 @@ def make_message(self, data):
     @staticmethod
     def validate_not_present(data, field_name):
         if field_name in data.keys():
+            logger.debug('Field cannon be set', field_name=field_name)
```
Comment (andrewmil) [inline app/validation/domain.py]: `Field cannot be set`?

File: app/application.py (hunk 8 above)
Comment (rjweeks) [inline app/application.py]: The user uuid gets logged during the authenticate call. But we will include the status in the log message.

File: app/application.py (hunk 9 above)
Comment (rjweeks) [inline app/application.py]: Ok we won't log the headers here.

File: app/authentication/authenticator.py (hunk 11 above)
Comment (rjweeks) [inline app/authentication/authenticator.py]: At the moment the other services do not encrypt the token.

**Pull request merged**